In [1]:
import pandas as pd

In [2]:
covid_df = pd.read_csv("italy-covid-daywise.csv")
covid_df.head()

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN


In [3]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47 entries, 0 to 46
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        47 non-null     object 
 1   new_cases   47 non-null     float64
 2   new_deaths  47 non-null     float64
 3   new_tests   30 non-null     float64
dtypes: float64(3), object(1)
memory usage: 1.6+ KB


In [4]:
covid_df.describe()

,new_cases,new_deaths,new_tests
count,47.000000,47.000000,30.000000
mean,2157.191489,295.468085,35284.433333
std,2065.001133,279.483285,14885.094583
min,-148.000000,0.000000,13665.000000
25%,537.500000,29.000000,26770.250000
50%,1401.000000,243.000000,32566.000000
75%,2875.000000,459.000000,39384.000000
max,6557.000000,971.000000,95273.000000


In [5]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [6]:
covid_df.shape

(47, 4)

In [7]:
import jovian

/home/geff/miniconda3/envs/ml_env/lib/python3.10/site-packages/jovian/utils/latest.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import parse_version


<IPython.core.display.Javascript object>

## Retrieving data from a data frame

In [8]:
# Pandas format is similar to this
covid_data_dict = {
    'date': ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases': [1444, 1365, 996,  975, 1326],
    'new_death': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

In [9]:
# Pandas format is not similar to this
covid_data_list = [
    {'date': '2020-09-01', 'new_cases': 996, 'new_deaths': 6, 'new_tests': None},
    {'date': '2020-09-02', 'new_cases': 996, 'new_deaths': 6, 'new_tests': 3564},
    {'date': '2020-09-03', 'new_cases': 996, 'new_deaths': 6, 'new_tests': 77459},
]

In [10]:
covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

In [11]:
covid_df['new_cases'].head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: new_cases, dtype: float64

In [12]:
type(covid_df['new_cases'])

pandas.core.series.Series

Just like arrays, you can retrieve a specific value with a series using the indexing
notation []

In [13]:
covid_df['new_cases'][45]

np.float64(975.0)

In [14]:
covid_df['new_tests'][26]

np.float64(22999.0)

Pandas also provides the .at method to directly retrieve data at a specific row and column

In [15]:
covid_df.at[35, 'new_cases']

np.float64(210.0)

In [16]:
covid_df.at[20, 'new_tests']

np.float64(24113.0)

In [17]:
type(covid_df.at[15, 'new_deaths'])

numpy.float64

Instead of using the indexing notation [], Pandas also allow accessing columns as properties
of the data frame using the `.` notation. However, this method only works for columns whose 
names do not contain special characters

In [18]:
covid_df.new_cases.head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: new_cases, dtype: float64

Further, you can also pass a list of columns within the indexing notation [] to access a subset of the
data frame with just the given columns

In [19]:
cases_df = covid_df[['date', 'new_cases']]
cases_df.head()

,date,new_cases
0,2019-12-31,0.0
1,2020-01-01,0.0
2,2020-01-02,0.0
3,2020-01-03,0.0
4,2020-01-04,0.0


Note, however that the new data frame `cases_df` is simply a 'view' of the original data frame `covid_df`.
Sometimes you might need a full copy of the data frame, in which case you use the `copy` method

In [20]:
covid_df_copy = covid_df.copy()

The data within `covid_df_copy` is completely separat from covid_df, and changing values inside one of them 
will not affect the other

To access a specific row of data, Pandas provides the `.loc` method

In [21]:
covid_df.loc[13]

date          2020-03-29
new_cases         5974.0
new_deaths         887.0
new_tests            NaN
Name: 13, dtype: object

Each retrieved row is also a Series object

In [22]:
type(covid_df.loc[13])

pandas.core.series.Series

In [23]:
covid_df.head(5)

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN


In [24]:
covid_df.tail(10)

,date,new_cases,new_deaths,new_tests
37,2020-06-19,331.0,66.0,28570.0
38,2020-06-20,-148.0,47.0,29875.0
39,2020-06-21,264.0,49.0,24581.0
40,2020-06-22,224.0,24.0,16152.0
41,2020-06-23,221.0,23.0,23225.0
42,2020-08-30,1444.0,1.0,53541.0
43,2020-08-31,1365.0,4.0,42583.0
44,2020-09-01,996.0,6.0,54395.0
45,2020-09-02,975.0,8.0,NaN
46,2020-09-03,1326.0,6.0,NaN


In [25]:
covid_df.at[0, 'new_tests']

np.float64(nan)

In [26]:
type(covid_df.at[0, 'new_tests'])

numpy.float64

The distinction between 0 and NaN is subtle but important. In this dataset, it represents that daily test numbers
were not reported on specific dates

We can find the first index that doesn't contain an NaN values using `first_valid_index` method

In [27]:
covid_df.new_tests.first_valid_index()

15

Let's look at a few rows before and after the index to verify that the values indeed change from NaN to actual
numbers. We can do this by passing a range to loc

In [28]:
covid_df.loc[13 : 16]

,date,new_cases,new_deaths,new_tests
13,2020-03-29,5974.0,887.0,NaN
14,2020-03-30,5217.0,758.0,NaN
15,2020-04-21,2256.0,454.0,28095.0
16,2020-04-22,2729.0,534.0,44248.0


The .sample method can be used to retrieve a random sample of rows from the data frame

In [29]:
covid_df.sample(10)

,date,new_cases,new_deaths,new_tests
13,2020-03-29,5974.0,887.0,NaN
1,2020-01-01,0.0,0.0,NaN
38,2020-06-20,-148.0,47.0,29875.0
25,2020-05-01,1900.0,474.0,27047.0
40,2020-06-22,224.0,24.0,16152.0
21,2020-04-27,2324.0,260.0,26678.0
20,2020-04-26,2357.0,415.0,24113.0
0,2019-12-31,0.0,0.0,NaN
22,2020-04-28,1739.0,333.0,37554.0
11,2020-03-27,6153.0,660.0,NaN


Notice that even though we have taken a random sample, the original index of each row has been preserve. This is an
important and useful property of data frames - each row of data has an index associated with it.

# Analyzing Data from data frames

Q: What is the total number of reported cases and deaths related to Covid-19 in Italy

Similar to Numpy arrays, a Pandas series supports the `sum` method to answer these questions

In [30]:
total_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()

print("The number of reported cases is {} and the number of reported deaths is {}".format(total_cases, total_deaths))

The number of reported cases is 101388.0 and the number of reported deaths is 13887.0


Q: What is the overall death rate (ratio of reported deaths to reported cases)

In [31]:
death_rate = covid_df.new_deaths.sum() / covid_df.new_cases.sum()

print("The overall reported death rate in Italy is {:.2f}%".format(death_rate))

The overall reported death rate in Italy is 0.14%


Q: What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were being reported

In [32]:
initial_tests = 935310
total_tests = initial_tests + covid_df.new_cases.sum()
total_tests

np.float64(1036698.0)

Q: What fraction of tests returned a positive result?

In [33]:
positive_rate = total_cases / total_deaths

print("{:.2f}% of tests in Italy led to a positive diagnosis.".format(positive_rate))

7.30% of tests in Italy led to a positive diagnosis.


# Querying and Sorting rows
Let's say we want only to look at the days which had more than 1000 reported cases

In [34]:
high_new_cases = covid_df.new_cases > 1000

In [35]:
high_new_cases.head()

0    False
1    False
2    False
3    False
4    False
Name: new_cases, dtype: bool

In [36]:
covid_df[high_new_cases].head()

,date,new_cases,new_deaths,new_tests
5,2020-03-20,5322.0,429.0,NaN
6,2020-03-21,5986.0,625.0,NaN
7,2020-03-22,6557.0,795.0,NaN
8,2020-03-23,5560.0,649.0,NaN
9,2020-03-25,5249.0,743.0,NaN


We can write this succintly on a single line by passing the boolean expression as an index to the data frame

In [37]:
high_cases_df = covid_df[covid_df.new_cases > 1000]
high_cases_df.head()

,date,new_cases,new_deaths,new_tests
5,2020-03-20,5322.0,429.0,NaN
6,2020-03-21,5986.0,625.0,NaN
7,2020-03-22,6557.0,795.0,NaN
8,2020-03-23,5560.0,649.0,NaN
9,2020-03-25,5249.0,743.0,NaN


We can also formulate more complex queries that involve multiple columns. As an example, let's try to determine the days when the ration of cases reported to tests conducted is higher than the overall `positive_rate`

In [38]:
positive_rate
high_ratio_df = covid_df[covid_df.new_cases / covid_df.new_tests > positive_rate]
high_ratio_df.head()

,date,new_cases,new_deaths,new_tests


Perfoming operations on multiple columns results in a new series

In [39]:
covid_df.new_cases / covid_df.new_tests

0          NaN
1          NaN
2          NaN
3          NaN
4          NaN
5          NaN
6          NaN
7          NaN
8          NaN
9          NaN
10         NaN
11         NaN
12         NaN
13         NaN
14         NaN
15    0.080299
16    0.061675
17    0.090877
18    0.027773
19    0.078110
20    0.097748
21    0.087113
22    0.046307
23    0.042806
24    0.062918
25    0.070248
26    0.060394
27    0.037906
28    0.028461
29    0.105671
30    0.030840
31    0.036768
32    0.034508
33    0.031044
34    0.018778
35    0.006184
36    0.009963
37    0.011586
38   -0.004954
39    0.010740
40    0.013868
41    0.009516
42    0.026970
43    0.032055
44    0.018311
45         NaN
46         NaN
dtype: float64

Further, we can use this series to add a new column to the data frame

In [40]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests
covid_df.head()

,date,new_cases,new_deaths,new_tests,positive_rate
0,2019-12-31,0.0,0.0,NaN,NaN
1,2020-01-01,0.0,0.0,NaN,NaN
2,2020-01-02,0.0,0.0,NaN,NaN
3,2020-01-03,0.0,0.0,NaN,NaN
4,2020-01-04,0.0,0.0,NaN,NaN


For now let's remove the `positive_rate` column using the  `drop` method

In [41]:
covid_df.drop(columns=['positive_rate'], inplace=True)

In [42]:
covid_df.head()

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN


# Sorting rows using column values

The rows can also be created by a specific column using `.sort_values`. Let's sort to identify the days with the highest number of cases, then chain it with the `head` method to get the 10 days with the most cases

In [43]:
# ascending=False sorts the values in decreasing order
covid_df.sort_values('new_cases', ascending=False).head(10)

,date,new_cases,new_deaths,new_tests
7,2020-03-22,6557.0,795.0,NaN
11,2020-03-27,6153.0,660.0,NaN
6,2020-03-21,5986.0,625.0,NaN
13,2020-03-29,5974.0,887.0,NaN
12,2020-03-28,5959.0,971.0,NaN
8,2020-03-23,5560.0,649.0,NaN
5,2020-03-20,5322.0,429.0,NaN
9,2020-03-25,5249.0,743.0,NaN
14,2020-03-30,5217.0,758.0,NaN
10,2020-03-26,5210.0,685.0,NaN


It looks like the last two weeks of March had the highest number of daily cases. Let's compare this to the days where the highest number of deaths were recorded

In [44]:
covid_df.sort_values('new_deaths', ascending=False).head(10)

,date,new_cases,new_deaths,new_tests
12,2020-03-28,5959.0,971.0,NaN
13,2020-03-29,5974.0,887.0,NaN
7,2020-03-22,6557.0,795.0,NaN
14,2020-03-30,5217.0,758.0,NaN
9,2020-03-25,5249.0,743.0,NaN
10,2020-03-26,5210.0,685.0,NaN
11,2020-03-27,6153.0,660.0,NaN
8,2020-03-23,5560.0,649.0,NaN
6,2020-03-21,5986.0,625.0,NaN
16,2020-04-22,2729.0,534.0,44248.0


Let's also look at the days with the least number of cases


In [45]:
# by default ascending=True meaning values will be sortede in ascending order
covid_df.sort_values('new_cases').head(10)

,date,new_cases,new_deaths,new_tests
38,2020-06-20,-148.0,47.0,29875.0
0,2019-12-31,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
35,2020-06-17,210.0,34.0,33957.0
41,2020-06-23,221.0,23.0,23225.0
40,2020-06-22,224.0,24.0,16152.0
39,2020-06-21,264.0,49.0,24581.0


Seems like the count of new cases on June 20th was -148, a negative number! Not something we might have expected, but that's the nature of real world data. It could simpy be a data entry error, or it's possible that the government may have issued to account for miscounting in the past

Let's look at some days before and after June 20th

In [46]:
covid_df.loc[36 : 40]

,date,new_cases,new_deaths,new_tests
36,2020-06-18,328.0,43.0,32921.0
37,2020-06-19,331.0,66.0,28570.0
38,2020-06-20,-148.0,47.0,29875.0
39,2020-06-21,264.0,49.0,24581.0
40,2020-06-22,224.0,24.0,16152.0


If this was indeed a data entry error, we can use one of the following approaches for dealing with the missing or faulty value
    1. Replace with 0
    2. Replace it with the average of the entire column
    3. Replace it with the average values on the previous and next date
    4. Discard the row entirely

The .at method can be used to modify a specific value within the data frame

In [47]:
covid_df.at[38, 'new_cases'] = (covid_df.at[37, 'new_cases'] + covid_df.at[39, 'new_cases']) / 2

In [48]:
covid_df.loc[36 : 40]

,date,new_cases,new_deaths,new_tests
36,2020-06-18,328.0,43.0,32921.0
37,2020-06-19,331.0,66.0,28570.0
38,2020-06-20,297.5,47.0,29875.0
39,2020-06-21,264.0,49.0,24581.0
40,2020-06-22,224.0,24.0,16152.0


# Working with dates

In [63]:
covid_df.date.head()

0   2019-12-31
1   2020-01-01
2   2020-01-02
3   2020-01-03
4   2020-01-04
Name: date, dtype: datetime64[ns]

The data type of date is currently `object`, so Pandas does not know that column is a date

We can convert it into a datetime column using the `pd.to_datetime` method

In [64]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [65]:
covid_df['date'].head()

0   2019-12-31
1   2020-01-01
2   2020-01-02
3   2020-01-03
4   2020-01-04
Name: date, dtype: datetime64[ns]

You can see that it now has the datatype `datetime64`. We can now extract different parts of the data into seperate columns, using the `DatetimeIndex` class

In [66]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [67]:
covid_df.head()

,date,new_cases,new_deaths,new_tests,year,month,day,weekday
0,2019-12-31,0.0,0.0,NaN,2019,12,31,1
1,2020-01-01,0.0,0.0,NaN,2020,1,1,2
2,2020-01-02,0.0,0.0,NaN,2020,1,2,3
3,2020-01-03,0.0,0.0,NaN,2020,1,3,4
4,2020-01-04,0.0,0.0,NaN,2020,1,4,5


Let's check the overall metrics for the month of May. We can query the rows for May, choose a subset of columns that we want to aggregate, and use the sum method of the data frame to get the sum of values in each choosen column

In [68]:
# Querying the rows for May
covid_df_may = covid_df[covid_df.month == 5]

In [69]:
covid_df_may.head()

,date,new_cases,new_deaths,new_tests,year,month,day,weekday
23,2020-05-01,1872.0,285.0,43732.0,2020,5,1,4
24,2020-05-01,1965.0,269.0,31231.0,2020,5,1,4
25,2020-05-01,1900.0,474.0,27047.0,2020,5,1,4
26,2020-05-01,1389.0,174.0,22999.0,2020,5,1,4
27,2020-05-01,1221.0,195.0,32211.0,2020,5,1,4


In [70]:
# Extract the subset of columns to be aggregated 
covid_df_may_metrics = covid_df_may[['new_cases', 'new_deaths', 'new_tests']]

In [71]:
covid_df_may_metrics.head()

,new_cases,new_deaths,new_tests
23,1872.0,285.0,43732.0
24,1965.0,269.0,31231.0
25,1900.0,474.0,27047.0
26,1389.0,174.0,22999.0
27,1221.0,195.0,32211.0


In [72]:
# Get the column-wise sum
covid_df_may_totals = covid_df_may_metrics.sum()
covid_df_may_totals

new_cases      16223.0
new_deaths      3057.0
new_tests     387013.0
dtype: float64

In [73]:
type(covid_df_may_totals)

pandas.core.series.Series

The operation above can also be combined into a single statement

In [74]:
covid_df[covid_df.month == 5][['new_cases', 'new_deaths', 'new_tests']].sum()

new_cases      16223.0
new_deaths      3057.0
new_tests     387013.0
dtype: float64

Here's another example, let's check if the number of cases reported on Sundays is higher than the average number of cases reported every day.

This time, we might want to aggregate using the .mean method

In [75]:
# Overall average
covid_df.new_cases.mean()

np.float64(2166.6702127659573)

In [76]:
# Average for Sundays
covid_df[covid_df.weekday == 6].new_cases.mean()

np.float64(3319.2)

It seems like more cases were reported on Sundays compared to other days

# Grouping and aggregation

AS a next step, we might want to summarize the daywise data and create a new dataframe with month-wise data

This is where the `groupby` function is useful. Along with a grouping, we need to specify a way to aggregate the data for each group

In [78]:
monthly_groups = covid_df.groupby('month')

In [79]:
monthly_groups[['new_cases', 'new_deaths', 'new_tests']].sum()

,new_cases,new_deaths,new_tests
month,,,
1,0.0,0.0,0.0
3,57187.0,7202.0,0.0
4,20442.0,3317.0,331720.0
5,16223.0,3057.0,387013.0
6,1875.5,286.0,189281.0
8,2809.0,5.0,96124.0
9,3297.0,20.0,54395.0
12,0.0,0.0,0.0


In [80]:
covid_month_df = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].sum()

In [82]:
covid_month_df

,new_cases,new_deaths,new_tests
month,,,
1,0.0,0.0,0.0
3,57187.0,7202.0,0.0
4,20442.0,3317.0,331720.0
5,16223.0,3057.0,387013.0
6,1875.5,286.0,189281.0
8,2809.0,5.0,96124.0
9,3297.0,20.0,54395.0
12,0.0,0.0,0.0


The result is a new data frame, which uses unique values from the column passed to `groupby` as the index. Grouping and aggregation is a really powerful method of progressively summarizing data into smaller data frames

Instead of aggregating by sum, you can also aggregate by other measures like mean

In [83]:
covid_weekday_mean_df = covid_df.groupby('weekday')[['new_cases', 'new_deaths', 'new_tests']].mean()

In [84]:
covid_weekday_mean_df

,new_cases,new_deaths,new_tests
weekday,,,
0,2938.000000,339.000000,28471.000000
1,1042.400000,163.200000,35817.250000
2,1832.600000,263.800000,39102.500000
3,2046.800000,234.200000,35002.000000
4,1804.411765,275.058824,36489.714286
5,3052.700000,412.600000,34275.500000
6,3319.200000,429.400000,34078.333333


Apart from grouping, another form of aggregation is to calculate the running or cumultative sum of cases, tests or deaths up to the current date for each row. This can be done using `cumsum` method.

In [86]:
covid_df['total_cases'] = covid_df.new_cases.cumsum()
covid_df['total_deaths'] = covid_df.new_deaths.cumsum()
covid_df['total_tests'] = covid_df.new_tests.cumsum() + initial_tests

We've also included the initial test count in total_tests to account fot test conducted before daily reporting was started

In [87]:
covid_df.head()

,date,new_cases,new_deaths,new_tests,year,month,day,weekday,total_cases,total_deaths,total_tests
0,2019-12-31,0.0,0.0,NaN,2019,12,31,1,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN,2020,1,1,2,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN,2020,1,2,3,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN,2020,1,3,4,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN,2020,1,4,5,0.0,0.0,NaN


Notice how the NaN values in the total_test column remain unaffected